src/backend/looped/weapons/shark_gun.cpp

```diff
@@ -0,0 +1,57 @@
+#include "backend/looped/looped.hpp"
+#include "core/enums.hpp"
+#include "gta/enums.hpp"
+#include "gui.hpp"
+#include "util/math.hpp"
+#include "util/vehicle.hpp"
+#include "util/ped.hpp"
+#include <numbers>
+
+namespace big
+{
+	static auto last_time = std::chrono::steady_clock::now();
+
+	inline Entity e_shark;
+	void looped::weapons_shark_gun()
+	{
+		const bool is_shark_gun_selected = g.weapons.custom_weapon == CustomWeapon::SHARK_GUN;
+
+		const auto time_now = std::chrono::steady_clock::now();
+
+		const auto elapsed_time_in_ms = std::chrono::duration_cast<std::chrono::milliseconds>(time_now - last_time);
+
+		if (is_shark_gun_selected && !g_gui->is_open() && elapsed_time_in_ms >= 100ms && PAD::IS_DISABLED_CONTROL_PRESSED(0, (int)ControllerInputs::INPUT_ATTACK)
+		    && (!g.self.custom_weapon_stop || WEAPON::IS_PED_ARMED(self::ped, 4 | 2)))
+		{
+				Vector3 location = self::pos;
+
+
+				constexpr int rotation_order = 2;
+
+				Vector3 rot = CAM::GET_GAMEPLAY_CAM_ROT(rotation_order);
+				float pitch = math::deg_to_rad(rot.x); // vertical
+				//float roll = rot.y;
+				float yaw = math::deg_to_rad(rot.z + 90); // horizontal
+
+				float dist = 10.f;
+				location.x += dist * cos(pitch) * cos(yaw);
+				location.y += dist * sin(yaw) * cos(pitch);
+				location.z += dist * sin(pitch);

```

Comment:
These calculations are wildly inefficient. 

Since rotation_order is a constant value (2), you can precompute the trigonometric functions for the fixed rotation order outside the loop (if applicable). This avoids redundant calculations during each iteration.

The expression cos(pitch) * cos(yaw) appears twice. You can compute it once and store it in a temporary variable to avoid redundant calculations.

Similarly, sin(yaw) * cos(pitch) can be computed once and reused. You also subsequently calling sine and cosine again and again after for velocity. You are basically putting forcing the compiler to generate awful code.

Reply:
Probably a remnant from my awful code in the past, this can indeed be fixed by assigning the return values into temporary variables.